### DISCUSSION TITLE
Cursordance?

### DISCUSSION THREAD

**Discussion by RichardKanshen** [ppy/osu#14635] [Ideas]
Okay, so, there are so many Cursordance bots and illegal programs that might lead to a ban. Most of them are according to YouTube videos pretty neat, but still, the risk of getting banned is quite high.

So, the question is...
**How could it be implemented?**
Basically, it could be implemented using on/off switch customization in Autoplay and Autopilot mods.

**Customizations**
_All of the lower stated customizations will unlock after switching on the cursordance option_

- Enable Cursordance skin - On enabling: Cursor middle will be glowing with colors slowly; Cursor trail will be longer and rainbow.
- Enable Spin Cursordance - 
- When disabled: Autoplay/Autopilot will complete spinners in ellipses randomly rotated; 
- When enabled: Autoplay/Autopilot will complete spinners in random patterns/shapes

**Summary**

So, I guess that this is everything I wanted to say in this Idea, big thanks to the osu!team for being awesome in osu!development :)

- Richard
:)

**Comment by SonoSooS** [comment]
I would actually implement the cursordance part if I could get it end up merged into the game.

As for the rest, I didn't read enough code yet to know how much the rest is possible, but even if possible, I would rather do it only one feature at a time, to increase my chances of my efforts being merged into the main codebase.

The last thing I want is spending weeks on a feature, just for it getting permanently rejected just because a feature "would not fit the game", or some other reason.


I just need a developer/maintainer who has merge access to approve this, so I could start working on it.

**Reply by bdach** [reply]
I wouldn't personally mind seeing a cursordance option in the game, but (a) that is just my opinion and I'm not sure everyone else will agree and (b) I _cannot_ guarantee that it is going to get merged in advance. If it's going to be PR'd and the code will suck then I won't approve of it no matter how cool it looks. That's just not how open source works in general.

Also out of curiosity have you actually developed a cursordance bot before? Doesn't look that way from your github profile.

**Reply by SonoSooS** [reply]
Due to complications, I have never published any of the work I have done with osu!.

But yes, I have actually implemented three autoplay engines (five if we count variations due to engine limitations).

Yes, I am very well aware that janky and obfuscated-looking code will never get merged, and I am very well aware of that.

I do have a lot of refactoring to do to port those 4-5years old autoplay code to lazer, as I have written those with absolutely no consistent code style. My code style has greatly improved, but even then I'd have to refactor my code style to match those of lazer's codebase anyway.

Edit: grammar

**Reply by PercyDan54** [reply]
I remeber there's already attempt to do this
https://github.com/TechnoJo4/osu/blob/master/osu.Game.Rulesets.Osu/Replays/OsuDanceAutoGenerator.cs

**Reply by SonoSooS** [reply]
I looked into it, but it seems to suffer from the same issues as my old implementation does.

The problem is is that this change "eats itself into the code", as opposed to the modular design which I am proposing.
In other words, what was linked could be considered as "fixed function", as it changes irrelevant pieces of the code to make it work as intended [[1]](https://github.com/TechnoJo4/osu/commit/5e0af9ba0809c5414f96950e79dd5f9f66fe7c37#diff-9616c7645fb081b2bc535d116470f70980924eef9fe5899433b7aa6aef03b598) [[2]](https://github.com/TechnoJo4/osu/blob/ddd7d8e870ccd8bc624869c12eb5afe23eb9e5f5/osu.Game.Rulesets.Osu/Configuration/OsuRulesetConfigManager.cs#L32).

My proposal would extend the UI of the configurator interface ("Customize" button in the mod selector popup), and you would configure the autoplay engine there. That way you won't need hacky solutions in the code, and this UI change could also be useful for future mod configurations as well.

Yes, the obvious downside with my solution is that you have to reconfigure the settings of the selected autoplay engine each time you enable it. But considering that lazer is not a hacked client, I don't think that autoplay configuration has to be so deeply embedded into the code.

But that is just my opinion, and also based on observations of the code style and design philosophy of the entire project.

**Reply by cdwcgt** [reply]
Going forward now that we have mod presets, configuration may not be an issue anymore.
And there is also an implementation available
https://github.com/MATRIX-feather/osu/blob/stable/osu.Game.Rulesets.Osu/Replays/OsuDanceAutoGenerator.cs

To me, as long as the configuration is decoupled from other code, it should be possible to implement this function

**Reply by peppy** [reply]
The main issue is bringing the code quality up to scratch.

**Reply by cdwcgt** [reply]
This design should be better?

each mover is a single mod and each mod and be customize
for now I only do flower one and need code cleanup

for sure the design can be accept
![image](https://github.com/ppy/osu/assets/34775378/3ed228d3-7fd4-4182-a76e-9faba4582bfe)

**Reply by Azekel9** [reply]
How do i use this

**Reply by bdach** [reply]
@Azekel9 it's not actually in the game yet

**Reply by peppy** [reply]
((it would still be cool to get this in))

**Reply by SonoSooS** [reply]
I still have the source code to mine from when I made this reply. I fully ported it to lazer, except for one feature, which relies on the circle size being constant through the entire map.

I got blocked by not being able to port a feature due to not being able to keep up the code style, and without that feature, it feels too simple (so, basically the normal autoplay, but with the interpolator changed), so I eventually forgot that I did this port.

What should I do?

Edit: I want to do a full rewrite of my engine (and probably should, it's old code from like 2018), so I can change the design by request. I don't think even the ported code from 2022 would work in current lazer anymore anyway. And that way, we could also hopefully narrow down maintenance to only a single class, as all of my autplay engines use a single base class, and parse the entire beatmap into a different representation, that could be interpreted more easily, and hopefully more portable this way.

Edit2: I found the old discussion (#18031). Do these requests/rules still stand?

**Reply by PercyDan54** [reply]
If anything I had [a fork](https://github.com/PercyDan54/osu/tree/custom/osu.Game.Rulesets.Osu/Replays/Danse) which ported [danser-go](https://github.com/Wieku/danser-go) to Lazer. The main problem is actually the config. 

I don't hope to get this merged because code quality issues:
Currently it reads from game config via static Instace field (it sucks I know), but putting all the settings to mod panel would be too many

**Reply by SonoSooS** [reply]
\*bump\* just in case the notification got eaten, my questions still stand. It's okay if the team is still thinking about it (please reply if so), just want to make sure that Microsoft didn't break notifications again.

**Reply by bdach** [reply]
@SonoSooS not sure what "questions" you're referring to specifically

**Reply by SonoSooS** [reply]
Those that I asked below peppy's `((it would still be cool to get this in))` reply.

Many years have elapsed since I originally ported my cursordance engine to lazer, and want to know if the requirements have changed since, should I or should I not touch existing autoplay infrastructure, etc.

Also should I try to get the existing cursordance merged, and add in the missing feature (that *currently* relies on fixed circle size) later?

Do we still want to implement it as the preset dropdown with no configurability? I've tried to minimize preset count while maximizing variety, but I don't know if many years later that's still adequate or not.

And if it's even a good or bad idea to have a fully rewritten autoplay generator engine that transmutes hit object data to Abstract Timing Event format, and then that gets used for the other autoplay engine(s). It's kind of necessary to do this transmutation due to the flexibility it offers to have the data represented in a different way, but due to subtle differences, it makes existing autoplays incompatible if we decide to use this engine as the Default.
Do we want two autoplay engines?


These are the most important details that I have extracted from my reply.

**Reply by bdach** [reply]
> Many years have elapsed since I originally ported my cursordance engine to lazer, and want to know if the requirements have changed since, should I or should I not touch existing autoplay infrastructure, etc.

Question too vague for reasonable answer. I still stand by most things I said in the previous thread, e.g. there should be no "settings that depend on other settings" stuff.

> Also should I try to get the existing cursordance merged, and add in the missing feature (that _currently_ relies on fixed circle size) later?

Again vague. What is the missing feature? Why does "fixed circle size" matter? We're not changing circle size mid gameplay yet, are we?

> Do we still want to implement it as the preset dropdown with no configurability? I've tried to minimize preset count while maximizing variety, but I don't know if many years later that's still adequate or not.

See answer 1 I guess.

> And if it's even a good or bad idea to have a fully rewritten autoplay generator engine that transmutes hit object data to Abstract Timing Event format, and then that gets used for the other autoplay engine(s). It's kind of necessary to do this transmutation due to the flexibility it offers to have the data represented in a different way, but due to subtle differences, it makes existing autoplays incompatible if we decide to use this engine as the Default. Do we want two autoplay engines?

Hearing "abstract timing event format" makes me scared. Maybe don't "abstract" anything just yet?

Do you have an actual branch I can look at or something so we don't talk vagueries in circles?

**Reply by SonoSooS** [reply]
It's not *too* vague, it's just that some context depends on the earlier discussions (that I have linked).

I do have a branch, but please don't judge the code quality itself, just the functionality, as it *has* to be rewritten. It's a direct port to lazer from an even older codebase with terrible code style, so it leaves a lot to be desired.

https://github.com/ppy/osu/compare/master...SonoSooS:osu:osu-18030-autoplay-alt

Edit: also keep in mind, that this is nowhere near ready for meging, this is still just porting the functionality, while also having some git history, just in case reverting is necessary

**Reply by peppy** [reply]
The only real advice I can give is to try and dissect it into small changes if you are changing the autoplay engine itself.

It might be better to keep it all in its own mod (if that's not already how it's done) to avoid having to modify the existing code at all.

**Reply by SonoSooS** [reply]
I think that adding this below "Cinema" and "Autoplay" as a separate mod should be the good idea then, and that should avoid quite lot of the issues I have mentioned (here and in the previous posts), and also shouldn't modify "critical" code then, hopefully.
